### PR TITLE
Add `Material#getPlacedType` API

### PR DESCRIPTION
**Pull request by tal5**
## Additions

- `Material#getPlacedType` - returns the type an item would be placed as (usually the same, but is different in some cases, e.g. pumpkin seeds -> pumpkin stem).
- `UnsafeValues#getPlacedType(Material)` - bridge method for `Material#getPlacedType`.

## Notes

- Currently it returns a `Material`, as that's what's stored internally - might make sense to return `BlockData` to avoid potential issues with more complex data being added in the future/as it might make sense to represent it as `BlockData`?
- `BlockData#getPlacementMaterial` (other way around from this) seems to return `AIR` when a block doesn't have an item equivalent, while this returns `null` for items that can't be placed.
`null` makes more sense imo, but either is fine I guess - let me know if that should be changed to match.

### CODE REVIEW

File: patches/api/0426-Add-Material-getPlacedType-API.patch [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
++     * <p>
++     * Some examples include:
++     * <pre>
++     *     {@link Material#REDSTONE} -> {@link Material#REDSTONE_WIRE}
++     *     {@link Material#PUMPKIN_SEEDS} -> {@link Material#PUMPKIN_STEM}
++     * </pre>
++     * @param item An item material.
++     * @return The material that would be placed down by {@code item}, or {@code null} if it can't be placed.
++     * @throws IllegalArgumentException if {@code item} isn't an item ({@link Material#isItem()} is {@code false})
++     */
```
Comment (Machine-Maker) [inline patches/api/0426-Add-Material-getPlacedType-API.patch:57]: Can drop the javadoc here, just increases workload to maintain it since its a copy

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
I think returning BlockData would be more appropriate if we had a method that called `BlockItem#getPlacementState` which returns the equivalent BlockData in nms (BlockState). For just this, Material is fine.

----

I guess since Material is just going to be straight up deprecated by upstream hopefully next major version, I don't really have an issue adding more methods to Material that only apply to items or blocks but not both

**Comment by Warriorrrr** [comment]
This seems to be a duplicate of https://github.com/PaperMC/Paper/pull/6680 if I'm not mistaken

**Comment by Owen1212055** [comment]
Closing, as in general we are waiting for upstream's blocktype/itemtype API to better expose this types of API.

**Pull request closed without merging**